[ci] Merge ESLint and Prettier workflows with auto-fix for faster iteration

snomiao opened the pull request:
## Summary
- Combines separate ESLint and Prettier GitHub Actions workflows into a unified `lint-and-format.yaml`
- Automatically fixes ESLint and Prettier issues on pull requests and commits changes back to the PR branch
- Adds necessary GitHub Action permissions (`contents: write`) to enable auto-commits
- Includes PR commenting to notify contributors when auto-fixes are applied

## Benefits
- **Faster iteration**: Eliminates the manual fix cycle where contributors need to run lint/format locally and push again
- **Reduced friction**: Contributors don't need to remember specific lint/format commands
- **Consistency**: Ensures all PRs automatically follow project coding standards
- **Developer experience**: Less back-and-forth between CI failures and manual fixes

## How it works
1. When a PR is opened, the workflow runs ESLint with `--fix` and Prettier with `--write`
2. If changes are made, they're automatically committed back to the PR branch
3. A comment is added to the PR notifying about the auto-fixes
4. Final lint and format checks run to ensure everything passes

This approach is commonly used in modern projects to streamline the development workflow and reduce CI/CD cycle times.

## Action Items After this branch Merged

- [ ] Update branch ruleset in setting, for [required checks]: remove [eslint] and [prettier], add [lint-and-format]
- [ ] Update Branch for ALL OTHER PULL REQUESTS

┆Issue is synchronized with this [Notion page](https://www.notion.so/PR-4638-ci-Merge-ESLint-and-Prettier-workflows-with-auto-fix-for-faster-iteration-2426d73d365081a09d00fe1f9d41c99a) by [Unito](https://www.unito.io)

snomiao committed:
[ci] Merge ESLint and Prettier workflows with auto-fix

- Combines separate eslint.yaml and format.yaml into unified lint-and-format.yaml
- Automatically fixes ESLint and Prettier issues on PRs and commits changes
- Adds GitHub Action permissions for content write to enable auto-commits
- Includes PR commenting to notify when auto-fixes are applied
- Enables faster iteration by eliminating manual fix cycles

snomiao committed:
chore(lint-and-format): remove final lint and format check steps to streamline workflow execution

snomiao committed:
Merge branch 'main' into sno-eslint-fix

christian-byrne commented on .github/workflows/lint-and-format.yaml:
```diff
@@ -0,0 +1,65 @@
+      - name: Checkout PR
+        uses: actions/checkout@v4
+        with:
+          token: ${{ secrets.GITHUB_TOKEN }}
```
1. Existing Pattern: At least 5+ workflows already use contents: write permission:
    - update-registry-types.yaml:15
    - update-manager-types.yaml:10
    - version-bump.yaml:22
    - update-electron-types.yaml
    - update-litegraph.yaml
  2. Standard Practice: 8+ workflows use peter-evans/create-pull-request or similar token patterns, indicating this is an established security model in the repo.
  3. Appropriate Scope: The permission is properly scoped to only the specific job that needs it, following GitHub Actions best practices.

christian-byrne commented on .github/workflows/lint-and-format.yaml:
```diff
@@ -0,0 +1,65 @@
+
+on:
+  pull_request:
+    branches-ignore: [ wip/*, draft/*, temp/* ]
```
This is just using the existing patterns

christian-byrne reviewed (commented):
## Review Feedback

@snomiao Two specific improvements needed:

### 1. Add Final Validation Step
After auto-fixes are applied, the workflow should run a final validation to ensure all issues are resolved:

```yaml
- name: Final validation
  if: steps.auto-fix.outputs.changes_made == 'true'
  run:  < /dev/null | 
    npm run lint
    npm run format:check
```

### 2. Improve PR Comment Message
The current auto-fix comment should warn about local/remote branch divergence:

```markdown
## 🔧 Auto-fixes Applied

This PR has been automatically updated to fix linting and formatting issues.

**⚠️ Important**: Your local branch is now behind. Run `git pull` before making additional changes to avoid conflicts.

### Changes made:
- ESLint auto-fixes  
- Prettier formatting
```

The security concern in the previous review is **not valid** - this repo already uses `contents: write` in 5+ workflows following the same pattern.

snomiao committed:
[chore] Improve lint-and-format workflow with final validation and enhanced PR comments

- Add final validation step to ensure lint and format checks pass after auto-fixes
- Enhance PR comment with clearer messaging and instructions for developers
- Fix YAML formatting in branches-ignore array for consistency

snomiao commented:
updated! @christian-byrne

snomiao commented:
> @snomiao how come the eslint and prettier jobs don't run on this branch itself?

@christian-byrne  because its merged from 2 to 1 and original workflow is removed

maybe a bypass merge is needed
and then we have to update branch rulesets（required checks） to lint+format

christian-byrne commented:
Oh yeah. I think we need to also update lines like these

https://github.com/Comfy-Org/ComfyUI_frontend/blob/e77411d9dab14fcfde5175c5683366d2fe385800/.github/workflows/claude-pr-review.yml#L25

snomiao committed:
[fix] Update claude-pr-review.yml to reference new lint-and-format job

christian-byrne reviewed (approved):
LGTM
